src/util-detect-file-hash.c

```diff
@@ -0,0 +1,377 @@
+ * \param line_no file line number from where the string was read
+ * \param type the hash algorithm
+ *
+ * \retval -1 failed to load the hash into the hash table
+ * \retval 1 successfully loaded the has into the hash table
+ */
+int LoadHashTable(ROHashTable *hash_table, char *string, char *filename,
+        int line_no, uint32_t type)
+{
+    /* allocate the maximum size a hash can have (in this case is SHA256, 32 bytes) */
+    uint8_t hash[32];
+    /* specify the actual size that should be read depending on the hash algorithm */
+    uint16_t size = 32;
+
+    if (type == DETECT_FILEMD5) {
+        size = 16;
+    }
+    else if (type == DETECT_FILESHA1) {
+        size = 20;
+    }
+
+    /* every byte represented with hexadecimal digits is two characters */
+    uint16_t expected_len = (size * 2);
+
+    if (ReadHashString(hash, string, filename, line_no, expected_len) == 1) {
+        if (ROHashInitQueueValue(hash_table, &hash, size) != 1)
+            return -1;
+    }
+
+    return 1;
+}
+
+/**
+ * \brief Match a hash stored in a hash table
+ *
+ * \param hash_table hash table that will hold the hash
+ * \param hash buffer containing the bytes of the has
+ * \param hash_len length of the hash buffer
+ *
+ * \retval 0 didn't find the specified hash
+ * \retval 1 the hash matched a stored value
+ */
+static int HashMatchHashTable(ROHashTable *hash_table, uint8_t *hash,
+        size_t hash_len)
+{
+    void *ptr = ROHashLookup(hash_table, hash, (uint16_t)hash_len);
+    if (ptr == NULL)
+        return 0;
+    else
+        return 1;
+}
+
+/**
+ * \brief Match the specified file hash
+ *
+ * \param t thread local vars
+ * \param det_ctx pattern matcher thread local data
+ * \param f *LOCKED* flow
+ * \param flags direction flags
+ * \param file file being inspected
+ * \param s signature being inspected
+ * \param m sigmatch that we will cast into DetectFileHashData
+ *
+ * \retval 0 no match
+ * \retval 1 match
+ */
+int DetectFileHashMatch (ThreadVars *t, DetectEngineThreadCtx *det_ctx,
+        Flow *f, uint8_t flags, File *file, Signature *s, SigMatch *m)
+{
+    SCEnter();
+    int ret = 0;
+    DetectFileHashData *filehash = (DetectFileHashData *)m->ctx;
+
+    if (file->txid < det_ctx->tx_id) {
+        SCReturnInt(0);
+    }
+
+    if (file->txid > det_ctx->tx_id) {
+        SCReturnInt(0);
+    }
+
+    if (file->state != FILE_STATE_CLOSED) {
+        SCReturnInt(0);
+    }
+
+    int match = -1;
+
+    if (s->file_flags & FILE_SIG_NEED_MD5 && file->flags & FILE_MD5) {
+        match = HashMatchHashTable(filehash->hash, file->md5, sizeof(file->md5));
+    }
+    else if (s->file_flags & FILE_SIG_NEED_SHA1 && file->flags & FILE_SHA1) {
+        match = HashMatchHashTable(filehash->hash, file->sha1, sizeof(file->sha1));
+    }
+    else if (s->file_flags & FILE_SIG_NEED_SHA256 && file->flags & FILE_SHA256) {
+        match = HashMatchHashTable(filehash->hash, file->sha256, sizeof(file->sha256));
+    }
+
```

Comment:
For those curious, the problem with #2102 was in the previous three conditions, where it wasn't checking to what hashing algorithm the signature refereed to.